vote/src/vote_account.rs

```diff
@@ -532,6 +586,78 @@ mod tests {
         assert!(vote_accounts.staked_nodes.get().unwrap().is_empty());
     }
 
+    #[test]
+    fn test_staked_nodes_update() {
+        let mut vote_accounts = VoteAccounts::default();
+
+        let mut rng = rand::thread_rng();
+        let pubkey = Pubkey::new_unique();
+        let node_pubkey = Pubkey::new_unique();
+        let (account1, _) = new_rand_vote_account(&mut rng, Some(node_pubkey));
+        let vote_account1 = VoteAccount::try_from(account1).unwrap();
+
+        // first insert
+        let ret = vote_accounts.insert(pubkey, vote_account1.clone(), || 42);
+        assert_eq!(ret, None);
+        assert_eq!(vote_accounts.get_delegated_stake(&pubkey), 42);
+        assert_eq!(vote_accounts.staked_nodes().get(&node_pubkey), Some(&42));
+
+        // update with unchanged state
+        let ret = vote_accounts.insert(pubkey, vote_account1.clone(), || {
+            panic!("should not be called")
+        });
+        assert_eq!(ret, Some(vote_account1.clone()));
+        assert_eq!(vote_accounts.get(&pubkey), Some(&vote_account1));
+        // stake is unchanged
+        assert_eq!(vote_accounts.get_delegated_stake(&pubkey), 42);
+        assert_eq!(vote_accounts.staked_nodes().get(&node_pubkey), Some(&42));
+
+        // update with changed state, same node pubkey
+        let (account2, _) = new_rand_vote_account(&mut rng, Some(node_pubkey));
+        let vote_account2 = VoteAccount::try_from(account2).unwrap();
+        let ret = vote_accounts.insert(pubkey, vote_account2.clone(), || {
+            panic!("should not be called")
+        });
+        assert_eq!(ret, Some(vote_account1.clone()));
+        assert_eq!(vote_accounts.get(&pubkey), Some(&vote_account2));
+        // stake is unchanged
+        assert_eq!(vote_accounts.get_delegated_stake(&pubkey), 42);
+        assert_eq!(vote_accounts.staked_nodes().get(&node_pubkey), Some(&42));
+
+        // update with new node pubkey, stake must be moved
+        let new_node_pubkey = Pubkey::new_unique();
+        let (account3, _) = new_rand_vote_account(&mut rng, Some(new_node_pubkey));
+        let vote_account3 = VoteAccount::try_from(account3).unwrap();
+        let ret = vote_accounts.insert(pubkey, vote_account3.clone(), || {
+            panic!("should not be called")
+        });
+        assert_eq!(ret, Some(vote_account2.clone()));
+        assert_eq!(vote_accounts.staked_nodes().get(&node_pubkey), None);
+        assert_eq!(
+            vote_accounts.staked_nodes().get(&new_node_pubkey),
+            Some(&42)
+        );
+    }
+
+    #[test]
+    fn test_staked_nodes_zero_stake() {
```

Comment:
I don't think this tests the zero stake bug completely. We should also have a case where `VoteAccounts` already has a vote account with zero stake, (so no staked node entry) and then update that vote account's node key to make sure that subtraction short circuits due to zero stake.

Reply:
added!